File: Binary_Search/LC_875_Koko_Eating_Bananas.py
# Written by: Jian-Yuan Lin
# Date: 2022.11.20

class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        l, r = 1, max(piles) + 1
        
        def cal_time(piles, speed):
            ans = 0
            for p in piles:
                ans += (p + speed - 1) // speed
            return ans
        
        while l < r:
            mid = l + (r-l) // 2
            time = cal_time(piles, mid)
            if time > h:
                l = mid + 1
            else:
                r = mid
        return l

File: Binary_Search/LC_875_Koko_Eating_Bananas.cpp
// Written by: Jian-Yuan Lin
// Date: 2022.11.20

class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int l = 1;
        int r = *max_element(piles.begin(), piles.end()) + 1;

        while (l < r){
            int mid = l + (r-l)/2;
            if (can_finish(piles, mid, h)){
                r = mid;
            }
            else{
                l = mid + 1;
            }
        }
        return l;
    }
private:
    bool can_finish(vector<int>& nums, int v, int h){
        int total = 0;
        for (int num: nums){
            total += (num + v - 1) / v;
        }
        if (total <= h){
            return true;
        }
        return false;
    }
};